core/src/banking_stage.rs

```diff
@@ -378,13 +383,257 @@ pub enum ForwardOption {
+                assert_eq!(
+                    buffered_packet_batches.packet_priority_queue.len(),
+                    buffered_packet_batches.message_hash_to_transaction.len()
+                );
+            }
+            Self::VoteStorage(latest_unprocessed_votes, VoteSource::Tpu) => {
+                // Based on the stake distribution present in the supplied bank, drain the unprocessed votes
+                // from each validator using a weighted random ordering. Votes from validators with
+                // 0 stake are ignored.
+                // TODO: Add proper error handling in case bank is missing?
+                let bank = bank.unwrap();
+
+                let latest_votes_per_pubkey = latest_unprocessed_votes
+                    .latest_votes_per_pubkey
+                    .read()
+                    .unwrap();
+
+                let retryable_votes = weighted_random_order_by_stake(&bank)
+                    .filter_map(|pubkey| {
+                        if let Some(lock) = latest_votes_per_pubkey.get(&pubkey) {
+                            if let Ok(latest_vote) = lock.write() {
+                                if let Ok(mut latest_vote) = latest_vote.try_borrow_mut() {
+                                    if !latest_vote.is_processed() {
+                                        let packet = latest_vote.clone();
+                                        latest_vote.clear();
+                                        latest_unprocessed_votes
+                                            .size
+                                            .fetch_sub(1, Ordering::AcqRel);
+                                        return Some(packet);
+                                    }
+                                }
+                            }
+                        }
+                        None
+                    })
+                    .chunks(batch_size)
+                    .into_iter()
+                    .flat_map(|vote_packets| {
+                        let vote_packets = vote_packets.into_iter().collect_vec();
+                        let packets_to_process = vote_packets
+                            .iter()
+                            .map(&DeserializedVotePacket::get_vote_packet)
+                            .collect_vec();
+                        if let Some(retryable_vote_indices) =
+                            processing_function(&packets_to_process)
+                        {
+                            retryable_vote_indices
+                                .iter()
+                                .map(|i| vote_packets[*i].clone())
+                                .collect_vec()
+                        } else {
+                            vote_packets
+                        }
+                    })
+                    .collect_vec();
+                // Insert the retryable votes back in
+                latest_unprocessed_votes.insert_batch(retryable_votes.into_iter());
```

Comment:
This `.clear()` into .insert_batch() is a bit cumbersome to maintain, especially because `insert_batch()` has a `assert!(!vote.is_processed());` which is a flag that might get cleared by the call to `clear()`

1. It seems like with this mechanism, once we've applied a vote to one fork, we can't apply it to another if we switch forks, which seems off.

```
  / --2
1
  \---3
```

I.e. here if the latest vote for a validator is 1, and we apply it to slot 2, but then switch to slot 3 as the major fork, we won't re-apply it to slot 3, whereas today we'll apply whatever `cluster_info_vote_listener` is sending us.

I think we should just apply whatever's in the bank on every slot for now, if it's already been in the bank we'll just fail with DuplicateSignature and shouldn't be included in the block since that's not a retryable error (can add a test for this for the VoteTransactions). In the future we can do something like what's in cluster_info_vote_listener and parse the vote states in the bank to figure out which votes to apply https://github.com/solana-labs/solana/blob/ed539d65b4c5851244c021d60c11844fb64c8d59/core/src/verified_vote_packets.rs#L123-L150

Reply:
> It seems like with this mechanism, once we've applied a vote to one fork, we can't apply it to another if we switch forks, which seems off.

I agree, although I think this behavior is already present, `UnprocessedPacketBatches` are cleared of the votes once successfully written to a bank. We rely on gossip or tpu to send us new votes when switching forks which is the same here.

Reply:
>Just apply whatever's in the bank on every slot for now, if it's already been in the bank we'll just fail with DuplicateSignature and shouldn't be included in the block since that's not a retryable error (can add a test for this for the VoteTransactions).

Is this not problematic as in every slot will now contain # staked validators votes even if the vote is a no-op.